File: src/writing-editor/docTypes.ts
/** Built-in document types — synced with `src-tauri/src/editor/doc_types.rs`. */

export type DocTypeCategory = "manuscript" | "planning";

export interface DocTypeDefinition {
  id: string;
  label: string;
  category: DocTypeCategory;
  icon: string;
  heading_level: number;
  builtin: boolean;
}

export const DOC_TYPES: DocTypeDefinition[] = [
  { id: "part", label: "Part", category: "manuscript", icon: "P", heading_level: 1, builtin: true },
  { id: "chapter", label: "Chapter", category: "manuscript", icon: "C", heading_level: 1, builtin: true },
  { id: "scene", label: "Scene", category: "manuscript", icon: "S", heading_level: 2, builtin: true },
  { id: "character", label: "Character", category: "planning", icon: "H", heading_level: 2, builtin: true },
  { id: "location", label: "Location", category: "planning", icon: "L", heading_level: 2, builtin: true },
  { id: "lore", label: "Lore", category: "planning", icon: "O", heading_level: 2, builtin: true },
  { id: "item", label: "Item", category: "planning", icon: "I", heading_level: 2, builtin: true },
  { id: "faction", label: "Faction", category: "planning", icon: "F", heading_level: 2, builtin: true },
  { id: "event", label: "Event", category: "planning", icon: "E", heading_level: 2, builtin: true },
  { id: "quest", label: "Quest", category: "planning", icon: "Q", heading_level: 2, builtin: true },
  { id: "note", label: "Note", category: "planning", icon: "N", heading_level: 3, builtin: true },
  { id: "timeline", label: "Timeline", category: "planning", icon: "T", heading_level: 2, builtin: true },
  { id: "theme", label: "Theme", category: "planning", icon: "M", heading_level: 3, builtin: true },
  { id: "folder", label: "Folder", category: "planning", icon: "D", heading_level: 1, builtin: true },
];

export function docTypesForSection(section: DocTypeCategory): DocTypeDefinition[] {
  return DOC_TYPES.filter((d) => d.category === section);
}

export function defaultDocTypeForSection(section: DocTypeCategory): string {
  return section === "manuscript" ? "chapter" : "note";
}

File: src-tauri/src/editor/doc_types.rs
//! Built-in document types for structured writing projects.

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct DocTypeDefinition {
    pub id: String,
    pub label: String,
    pub category: String,
    pub icon: String,
    pub heading_level: u8,
    pub builtin: bool,
}

pub fn builtin_doc_types() -> Vec<DocTypeDefinition> {
    vec![
        manuscript("part", "Part", "P", 1),
        manuscript("chapter", "Chapter", "C", 1),
        manuscript("scene", "Scene", "S", 2),
        planning("character", "Character", "H", 2),
        planning("location", "Location", "L", 2),
        planning("lore", "Lore", "O", 2),
        planning("item", "Item", "I", 2),
        planning("faction", "Faction", "F", 2),
        planning("event", "Event", "E", 2),
        planning("quest", "Quest", "Q", 2),
        planning("note", "Note", "N", 3),
        planning("timeline", "Timeline", "T", 2),
        planning("theme", "Theme", "M", 3),
        planning("folder", "Folder", "D", 1),
    ]
}

fn manuscript(id: &str, label: &str, icon: &str, heading_level: u8) -> DocTypeDefinition {
    DocTypeDefinition {
        id: id.into(),
        label: label.into(),
        category: "manuscript".into(),
        icon: icon.into(),
        heading_level,
        builtin: true,
    }
}

fn planning(id: &str, label: &str, icon: &str, heading_level: u8) -> DocTypeDefinition {
    DocTypeDefinition {
        id: id.into(),
        label: label.into(),
        category: "planning".into(),
        icon: icon.into(),
        heading_level,
        builtin: true,
    }
}

pub fn doc_type_category(doc_type: &str) -> Option<&'static str> {
    match doc_type {
        "part" | "chapter" | "scene" => Some("manuscript"),
        "character" | "location" | "lore" | "item" | "faction" | "event" | "quest" | "note"
        | "timeline" | "theme" | "folder" => Some("planning"),
        _ => None,
    }
}

pub fn is_known_doc_type(doc_type: &str) -> bool {
    doc_type_category(doc_type).is_some()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn has_fourteen_builtin_types() {
        assert_eq!(builtin_doc_types().len(), 14);
    }

    #[test]
    fn scene_is_manuscript_category() {
        assert_eq!(doc_type_category("scene"), Some("manuscript"));
        assert_eq!(doc_type_category("character"), Some("planning"));
    }
}
